app/models/mailing_list/signup.rb

```diff
@@ -0,0 +1,195 @@
+        "preferred_teaching_subject_id" => preferred_teaching_subject_id,
+        "accepted_policy_id" => accepted_policy_id,
+        "accept_privacy_policy" => accept_privacy_policy,
+        "qualification_id" => qualification_id,
+        "candidate_id" => candidate_id,
+        "address_postcode" => address_postcode,
+      }.compact
+    end
+
+    def latest_privacy_policy
+      @latest_privacy_policy ||= GetIntoTeachingApiClient::PrivacyPoliciesApi.new.get_latest_privacy_policy
+    end
+
+    def add_member_to_mailing_list!
+      request = GetIntoTeachingApiClient::MailingListAddMember.new(export_camelized_hash)
+      Rails.logger.info("MailingList::Signup#add_mailing_list_member!: #{::AttributeFilter.filtered_json(request)}")
+      api.add_mailing_list_member(request)
+    end
+
+    def already_signed_up?
+      request = GetIntoTeachingApiClient::ExistingCandidateRequest.new(identity_data)
+      GetIntoTeachingApiClient::CandidatesApi.new.create_candidate_access_token(request)
+
+      true
+    rescue GetIntoTeachingApiClient::ApiError => e
+      raise if e.code == 429
+
+      Rails.logger.info("#{self.class} potential duplicate (response code #{e.code})") unless e.code == 404
+
+      false
+    end
+
+  private
+
+    def identity_data
+      {
+        "email" => email,
+        "firstName" => first_name,
+        "lastName" => last_name,
+      }
+    end
+
+    def query_degree_status
+      GetIntoTeachingApiClient::PickListItemsApi.new.get_qualification_degree_status
+    end
+
+    def query_teaching_subjects
+      GetIntoTeachingApiClient::LookupItemsApi.new.get_teaching_subjects.reject do |type|
+        GetIntoTeachingApiClient::Constants::IGNORED_PREFERRED_TEACHING_SUBJECTS.values.include?(type.id)
+      end
+    end
+
+    def query_consideration_journey_stages
+      GetIntoTeachingApiClient::PickListItemsApi.new.get_candidate_journey_stages
+    end
+
+    def consideration_journey_stage_ids
+      consideration_journey_stages.map { |option| option.id.to_i }
+    end
+
+    def teaching_subject_ids
+      teaching_subjects.map(&:id)
+    end
+
+    def degree_status_option_ids
+      degree_status_options.map { |option| option.id.to_i }
+    end
+
+    def channel_ids
+      query_channels.map { |channel| channel.id.to_i }
+    end
+
+    def strip_whitespace
+      email&.strip!
+      first_name&.strip!
+      last_name&.strip!
+      timed_one_time_password&.strip!
+      address_postcode&.strip!
+    end
+
+    def api
+      GetIntoTeachingApiClient::MailingListApi.new
+    end
+
+    def timed_one_time_password_is_correct
+      request = GetIntoTeachingApiClient::ExistingCandidateRequest.new(identity_data)
+      response = api.exchange_access_token_for_mailing_list_add_member(timed_one_time_password, request)
+
+      Rails.logger.info("MailingList::Wizard#exchange_access_token: #{AttributeFilter.filtered_json(response)}")
+
+      response.tap do |r|
```

Comment:
I don't feel strongly about it, but this method is doing quite a lot and more than the method name suggests. I wonder if it would be clearer to split this and have an `after_validation :prepopulate` that sets the necessary ids etc from the totp response? You then have to track the response as an instance variable, so maybe what you have is more pragmatic 🤷‍♂️

Reply:
Yeah I did try with a separate method but the logic became a bit hard to follow. I'll rename the method to add a bit more clarity